src/Chainweb/Miner/Core.hs

```diff
@@ -107,26 +108,29 @@ mine
   -> Nonce
   -> TargetBytes
   -> HeaderBytes
-  -> IO HeaderBytes
-mine _ nonce (TargetBytes tbytes) (HeaderBytes hbytes) = BA.withByteArray tbytes $ \trgPtr -> do
-    !ctx <- hashMutableInit @a
-    fmap HeaderBytes . BA.copy hbytes $ \buf ->
-        allocaBytes (powSize :: Int) $ \pow -> do
-
-            -- inner mining loop
-            --
-            let go !n = do
-                    -- Compute POW hash for the nonce
-                    injectNonce n buf
-                    hash ctx buf pow
-
-                    -- check whether the nonce meets the target
-                    fastCheckTarget trgPtr (castPtr pow) >>= \case
-                        True -> pure ()
-                        False -> go (succ n)
-
-            -- Start inner mining loop
-            go nonce
+  -> IO (T2 HeaderBytes Word64)
+mine _ orig@(Nonce o) (TargetBytes tbytes) (HeaderBytes hbytes) = do
+    nonces <- newIORef 0
+    BA.withByteArray tbytes $ \trgPtr -> do
+      !ctx <- hashMutableInit @a
+      new <- fmap HeaderBytes . BA.copy hbytes $ \buf ->
+          allocaBytes (powSize :: Int) $ \pow -> do
+
+              -- inner mining loop
+              --
+              let go !n@(Nonce nv) = do
+                      -- Compute POW hash for the nonce
+                      injectNonce n buf
+                      hash ctx buf pow
+
+                      -- check whether the nonce meets the target
+                      fastCheckTarget trgPtr (castPtr pow) >>= \case
+                          True -> writeIORef nonces (nv - o)
+                          False -> go (succ n)
+
+              -- Start inner mining loop
+              go orig
+      T2 new <$> readIORef nonces
```

Comment:
Note for the future: this code has been heavily profiled and is not to be altered without confirming that no performance regression has been introduced.

Reply:
Dropping the stats printing from this loop makes the stats night unto worthless because in the real world mining blocks is quite a rare event.  We need to figure out a way to report results at regular intervals.